js-pkg/apps/demos/src/components/webrtc.js

```
@@ -0,0 +1,139 @@
+      } catch (e) {
+        console.error(e)
+      } finally {
+        makingOfferRef.current = false
+      }
+    }
+
+    conn.onicecandidate = (e) => {
+      let candidate = e.candidate
+      setSignalingMessages({ type: 'candidate', candidate })
+    }
+
+    conn.onconnectionstatechange = () => {
+      if (conn.connectionState === 'connected') {
+        setIsReady(true)
+      } else {
+        setIsReady(false)
+      }
+    }
+
+    conn.oniceconnectionstatechange = (_e) => {
+      if (conn.iceConnectionState === 'failed' || conn.iceConnectionState === 'disconnected') {
+        console.log('conn failed, restarting ICE')
+        conn.restartIce()
+      }
+    }
+
+    connRef.current = conn
+
+    for (let func of connSetupArray) {
+      func(conn)
+    }
+
+    return () => {
+      conn.close()
+    }
+  }, [connSetupArray])
+
+  React.useEffect(() => {
+    ;(async () => {
+      let conn = connRef.current
+      let makingOffer = makingOfferRef.current
+      const offerCollides = makingOffer || conn.signalingState !== 'stable'
+      const ignoreOffer = !isPeerPolite && offerCollides
+      let processed = 0
+      try {
+        for (let msg of signalingMessages) {
+          console.log(msg)
+          processed++
+          switch (msg.type) {
+            case 'answer':
+              await conn.setRemoteDescription(msg)
+              break
+            case 'offer':
+              if (ignoreOffer) break
+              await conn.setRemoteDescription(msg)
+              await conn.setLocalDescription()
+              setSignalingMessages({ type: 'answer', sdp: conn.localDescription.sdp })
+              break
+            case 'candidate':
+              try {
+                await conn.addIceCandidate(msg.candidate)
+              } catch (err) {
+                if (!ignoreOffer) throw err
+              }
+              break
+            default:
+              throw 'unknown message'
```

You should throw an actual error here instead of just a string: `throw new Error('unknown message')`